### ISSUE TITLE
RF annotation can impact model parameterization even without --hand

### ISSUE THREAD

**Issue by nawrockie** [EddyRivasLab/hmmer#180]
Due to how esl_msaweight_PB{_adv}() is implemented, RF annotation can impact MSA weighting if the provided RF annotation is not identical to what esl_msaweight_PB_adv()'s method for determining RF (match/insert) would determine. This means the same input alignment to hmmbuild with and without RF annotation can result in different models. I'm not sure if this is undesirable in hmmer3 or not, but it was in Infernal where RF should be ignored unless --hand is used. The fix in Infernal was simple: (https://github.com/EddyRivasLab/infernal/commit/fba203f88af0c42a5d91ab8c1cc932f45e8c79ea)

**Comment by cryptogenomicon** [comment]
I did that deliberately, actually! Did you come across a case where it was a bad thing to do?

My thinking was that if a reference annotation line is already provided, then it's good enough to use it to define consensus columns for sequence weighting. It's not like sequence weighting is guaranteed to use exactly the same set of consensus columns as model building anyway - because when you get to model building, you have seq weights, and they're used to calculate per-column residue-vs-gap occupancy.

**Comment by nawrockie** [comment]
I'm not surprised that it was deliberate, and I take your point about
sequence weighting not being guaranteed to set
consensus columns the same way as model building, but what is
undesirable from my perspective is that the RF annotation influences
the model building by default (when --hand is not used), when
previously this was not the case. Previously, a simplifying guarantee
in Infernal was that the same alignment (same sequences aligned the
same way, ignoring RF) gives the same model in cmbuild (if all
cmbuild options are consistent). I created this hmmer issue in case
you feel the same way about hmmer. I'd like to keep this assumption
in Infernal moving forward. There is one case (described below) where
it directly affects the Rfam family building pipeline. 



I remember that we (you and I) made an explicit choice to prior to the
infernal 1.1 release to have --hand be the option to use if cmbuild
should consider the RF annotation, otherwise it should be ignored.

Imagine a couple of (admittedly contrived) cases that I think mildly
argue for sequence weighting being agnostic of RF annotation in
cmbuild (which of course apply to hmmbuild as well):

1. A user builds model 1 with cmbuild from alignment 1 and saves the
   annotated alignment with the -O option as alignment 2. A possible
   model 2 built with alignment 2 may differ from model 1 if msa
   weighting is not RF agnostic.

2. A user takes a cmbuild or cmalign output alignment with RF
   annotation and removes some sequences but does not change the RF
   annotation creating alignment 1. Another user does the same thing
   (removes the same sequences) but also removes the RF annotation to
   create alignment 2. Alignments 1 and 2 are identical except for the
   RF annotation. Models built from alignments 1 and 2 may be
   different if msa weighting is not RF agnostic.

The case in the Rfam family building pipeline that caused a problem
for me is convoluted and difficult to explain, but here goes:

The Rfam team determined that when a new model is created the input
SEED alignment (potentially submitted by an expert on the RNA family
(think Tom)) should be unchanged, even with respect to the relative
position of nucleotides in insert columns, which cmbuild doesn't care
about. (Some users complained about this when their submitted SEED
changes in this way in the Rfam database.) However, we also wanted to
add RF annotation on that SEED so the Rfam user knows which positions
are consensus/insert. The obvious way to do the latter is to use
cmbuild -O, but then the inserts are left/right justified potentially
differently than the input SEED. So the problem is how to then put the
inserts back in their original positions.

To handle this, the Rfam script builds a CM (call it CM1) from the
original SEED (SEED1) and then maps the RF annotation from the CM1 -O
output alignment onto SEED1 to create SEED2, which is SEED1 with the
cmbuild RF annotation. SEED2 and CM1 are now stored in Rfam (and
downloadable as part of Rfam.seed and Rfam.cm).

This all works fine as of infernal 1.1.2 (prior to non-agnostic msa
weighting which came in with easel in infernal 1.1.3), but I think
that you would agree that it should follow that a CM2 model built from
SEED2 should then be identical to CM1. That is, if a user downloads
the SEED from Rfam and uses cmbuild to build a model (CM2) it should
be identical to the model stored in Rfam.cm (CM1).

However, with infernal 1.1.3 this is not guaranteed because CM2 will
potentially have different msa weights than CM1 did due to the
prescence of the RF annotation. You could say that, 'ok, then store
CM2 in Rfam', but that doesn't work either because it's possible that
the RF annotation from CM2 could be different than it was for CM1 due
to the different msa weights. So it's impossible to guarantee that we
can even create a SEED with RF annotation using this method that will
then result in the corresponding CM.

There's probably a workaround possible in the Rfam family pipeline to
deal with the non-agnostic msa weighting, but it would require
reworking this method. I considered doing that, but in my mind a
simpler and better solution is to bring back the guarantee that the
same alignment (same sequences aligned the same way) gives the same
model, regardless of whether RF annotation exists or not. If --hand is
used, it's a different story because then RF annotation is set already
and the issue referred to above is avoided.

**Comment by nawrockie** [comment]
I've thought of a more straightforward example of when RF-dependent msaweighting is a bad thing to do, in my opinion. 

1. build model M1 from alignment file A (without RF) and save the hmmbuild/cmbuild annotated alignment with -O to OA (with RF)
2. build model M2 from alignment file OA

I think the expectation would be that M1 and M2 are identical (do you agree?). However they may not be because msa weighting used to build M1 may differ from MSA weighting used to build model M2 due to the presence of RF in OA and absence of RF in A.

**Comment by cryptogenomicon** [comment]
Sorry to be slow getting to this. I agree with you. I've just checked in modifications in HMMER3 corresponding to your changes in Infernal.  

I think this is the right way to handle it, as opposed to changing the default behavior in Easel's esl_msaweight_PB(). Easel sees an MSA with vs. without RF annotation as two different inputs - providing an optional RF annotation in the input MSA is the moral equivalent of setting an optional command-line flag -  so it's ok to get two slightly different sets of weights. And using RF annotation is faster, saving Easel PB weighting from having to determine its own consensus cols. But you're right about how we want HMMER and Infernal to behave when they use PB weights.

Closing - thanks!
